[14.0] [NEW] sale_product_approval: New module added in the repo.

Chandresh-OSI opened the pull request:
CC @patrickrwilson @dreispt

patrickrwilson commented:
@Chandresh-OSI I was testing this on my local instance and couldn't get the 'sale_ok' field to properly set when changing stages. During troubleshooting, if I made the sale_ok not readonly it worked so that seemed to be preventing the field to change properly. I'm testing on enterprise, can you try this on enterprise and see if you see the same issue?

dreispt commented on sale_product_approval/models/product_state.py:
```diff
@@ -0,0 +1,22 @@
+    _inherit = "product.state"
+
+    approved_sale = fields.Boolean(
+        string="Approved to be Sold",
```
To not break current behavior once installed 
```suggestion
        string="Approved to be Sold",
        default=True,
```

dreispt commented on sale_product_approval/models/product_product.py:
```diff
@@ -0,0 +1,48 @@
+    sale_ok = fields.Boolean("Can be Sold", default=False)
+    candidate_sale = fields.Boolean(
+        string="Candidate to be Sold",
+    )
```
On install, we need a script to copy the `sale_ok` values to the `candidate_sale` column.

dreispt commented on sale_product_approval/models/product_product.py:
```diff
@@ -0,0 +1,48 @@
+    )
+
+    @api.onchange("product_state_id")
+    def _onchange_product_state(self):
```
This won't work for direct API writes, only for user form interaction.
I suggest instead adding a new-stype "onchange" computation to sale_ok.

dreispt commented on sale_product_approval/models/product_product.py:
```diff
@@ -0,0 +1,48 @@
+    def _onchange_product_state(self):
+        self.sale_ok = False
+        if self.product_state_id.approved_sale and self.candidate_sale:
+            self.sale_ok = True
```
```suggestion
        if self.product_state_id:
            self.sale_ok = self.candidate_sale and self.product_state_id.approved_sale
```

dreispt commented on sale_product_approval/models/product_product.py:
```diff
@@ -0,0 +1,48 @@
+            self.sale_ok = True
+
+    @api.model
+    def _get_default_product_state_id(self):
```
Is this missing from the product_state module?

dreispt commented on sale_product_approval/models/product_product.py:
```diff
@@ -0,0 +1,48 @@
+                )
+                .mapped("order_id")
+            )
+            self.env["sale.order"]._log_product_state(order_ids, self)
```
Instead of "pushing" these updates to Sales Orders, we could have them "pulled" on Sales Orders Lines with a computed field, and then on Sales Order with a computed field depending on SO Lines.

dreispt commented on sale_product_approval/models/product_product.py:
```diff
@@ -0,0 +1,48 @@
+        return res
+
+    @api.onchange("product_state_id")
+    def _onchange_product_state(self):
```
Not sure we will need this, once we have a computed field.

dreispt commented on sale_product_approval/models/product_state.py:
```diff
@@ -0,0 +1,22 @@
+                [("product_state_id", "=", self.id), ("candidate_sale", "=", True)]
+            )
+            for product in product_ids:
+                product.sale_ok = True
```
Make sale_ok computed, and this won't be needed.

dreispt commented on sale_product_approval/models/sale_order.py:
```diff
@@ -0,0 +1,52 @@
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    exceptions = fields.Boolean(
```
Field name look too short prone to name collisions 
```suggestion
    exceptions_sale_approval = fields.Boolean(
```

dreispt commented on sale_product_approval/models/sale_order.py:
```diff
@@ -0,0 +1,52 @@
+        if any(not line.approved_sale for line in self.order_line):
+            self.exceptions = True
+
+    def _log_exception_activity_sale(self, render_method, documents, product_id):
+        for order in documents:
```
OOP antipattern
```suggestion
    def _log_exception_activity_sale(self, render_method, product_id):
        for order in self:
```

dreispt commented on sale_product_approval/models/sale_order.py:
```diff
@@ -0,0 +1,52 @@
+                user_id=order.user_id.id or SUPERUSER_ID,
+            )
+
+    def _log_product_state(self, documents, product_id, cancel=False):
```
I now see why there we have the write() logic calling this.
Thinking that maybe we could do this with a cron, but let's not go that way for now.

dreispt replied:
Still not fixed.

patrickrwilson commented:
@Chandresh-OSI It mostly works, I can see the exception banner on the SO and the icon on the SO line but the exception activity doesn't seem to be created anymore. Is this on purpose or a bug?
![image](https://user-images.githubusercontent.com/36892066/126185202-46fc696a-f959-46a9-a538-5438f6216283.png)

patrickrwilson commented:
@Chandresh-OSI Tested and works great after the last commit, just need a bit more coverage on tests.

dreispt commented on sale_product_approval/models/product_product.py:
```diff
@@ -0,0 +1,50 @@
+        default=True,
+    )
+
+    @api.depends("product_state_id")
```
Missing dependencies
```suggestion
    @api.depends("candidate_sale", "product_state_id.approved_sale")
```

dreispt commented on sale_product_approval/models/product_product.py:
```diff
@@ -0,0 +1,50 @@
+                    product_variant_ids = (
+                        self.env["product.product"]
+                        .search([("product_tmpl_id", "=", product.id)])
+                        .ids
```
Tip: as a general rule, work with records, and extracts ids as late as possible

dreispt commented on sale_product_approval/models/product_product.py:
```diff
@@ -0,0 +1,50 @@
+                        self.env["sale.order.line"]
+                        .search(
+                            [
+                                ("product_id", "in", product_variant_ids),
```
`product_variant_ids` is available as a One2many field
```suggestion
                                ("product_id", "in", product.product_variant_ids.ids),
```

dreispt commented on sale_product_approval/models/product_state.py:
```diff
@@ -0,0 +1,23 @@
+                    product.sale_ok = True
+                else:
+                    product.sale_ok = False
+        return res
```
`sale_ok` is already computed, so this code should not be needed.
```suggestion
```

dreispt commented on sale_product_approval/models/sale_order.py:
```diff
@@ -0,0 +1,65 @@
+        if any(not line.approved_sale for line in self.order_line):
+            self.exceptions_sale_approval = True
+
+    def _log_exception_activity_sale(self, render_method, documents, product_id):
```
OOP smells: `documents` is a Sales Order recordset. Just use `self`!

dreispt commented on sale_product_approval/models/sale_order.py:
```diff
@@ -0,0 +1,65 @@
+        string="Exception",
+    )
+
+    def _compute_exceptions(self):
```
Missing dependencies
```suggestion
    @api.depends("order_line.approved_sale")
    def _compute_exceptions(self):
```

dreispt commented on sale_product_approval/models/sale_order.py:
```diff
@@ -0,0 +1,65 @@
+    def _compute_exceptions(self):
+        self.exceptions_sale_approval = False
+        if any(not line.approved_sale for line in self.order_line):
+            self.exceptions_sale_approval = True
```
Simplifying
```suggestion
        self.exceptions_sale_approval = any(not line.approved_sale for line in self.order_line)
```

dreispt commented on sale_product_approval/models/product_product.py:
```diff
@@ -0,0 +1,50 @@
+                        )
+                        .mapped("order_id")
+                    )
+                    self.env["sale.order"]._log_product_state(order_ids, self)
```
Can we trigger the exception directly on the the Sale order Line, and have that logic there propagate the exception status up to the Sales Order?
The _log_product_state method causes an exception to be presented so the method name is surprising, and that's not a good thing.

dreispt commented on sale_product_approval/models/sale_order.py:
```diff
@@ -0,0 +1,65 @@
+            )
+
+    def _log_product_state(self, documents, product_id, cancel=False):
+        def _render_product_state_excep(order, product_id):
```
Why a nested method. These are not extensible, do we want to prevent that?

dreispt commented on sale_product_approval/models/sale_order.py:
```diff
@@ -0,0 +1,65 @@
+                values=values
+            )
+
+        self.env["sale.order"]._log_exception_activity_sale(
```
Why `self.env["sale.order"]`if self is already a a sales order?

dreispt commented on sale_product_approval/models/sale_order.py:
```diff
@@ -0,0 +1,65 @@
+            )
+
+        self.env["sale.order"]._log_exception_activity_sale(
+            _render_product_state_excep, documents, product_id
```
Why pass a method as an argument?
This adds complexity, and I don''t see the need or the benefit.

dreispt commented:
@Chandresh-OSI @patrickrwilson There are some coding issues here that need improvement, can you work on this please?

Also, there is already a feature for exception handling.
It should be leveraged, and improved if it doesn't do the Activities already.
See https://github.com/OCA/sale-workflow/tree/14.0/sale_exception

patrickrwilson commented:
@Chandresh-OSI Looks very close but can you check travis? A test may need changes resulting from your last update. Thanks.

patrickrwilson commented:
@Chandresh-OSI can you squash commits then this should be ready to move forward.

Chandresh-OSI commented:
supersede by #1713

dreispt commented:
Why a new PR was needed?

Chandresh-OSI commented:
Due to commit squash issue. I have some merge commits in this PR and it is hard to squash individual commits. So I created new PR for the same code.

dreispt commented:
That doesn't require a new PR that restarts the review process.
You squash commits with `git rebase -i` and then force push the same branch.
Do you want to reopen and try that?
Reach out to me if you need more details.

Chandresh-OSI committed:
[ADD] sale_product_approval: Added new module in repo.

[IMP] Improved the code and added new field to override exception.

[IMP] allow candidate fields visible in debug mode

Users can now go into debug mode to see the candidate fields but this also adds compute to prevent users who do not have product admin group to be able to change the candidate field.

pre-commit

[FIX] sale_product_approval: do not hide Sales smart button, avoid install errors

[FIX] sale_product_approval: fix tests

[IMP] Improved code with lazy 'Can be' flags

[FIX] fixed pre-commit

[IMP] Added 'Can be added to SO'

Sometimes a product can be quoted even before it's ready to be sold or a product can stop being added to SO's but still allow SO to confirm, so this adds that option.

Reviewer Suggestions

patrickrwilson reviewed (approved):
This works great, LGTM

dreispt commented:
...there seems to be an issue with `sale_order_carrier_auto_assign` that is causing tests to fail: https://app.travis-ci.com/github/OCA/sale-workflow/jobs/536164726#L2357

dreispt commented:
Trying merge anyway...

dreispt commented:
/ocabot merge nobump

rousseldenis commented on sale_product_approval/models/product_product.py:
```diff
@@ -0,0 +1,46 @@
+                    product.candidate_sale and product.product_state_id.approved_sale
+                )
+                if not product.sale_ok:
+                    order_ids = (
```
Try to move this search outside the loop for performances reasons

rousseldenis commented on sale_product_approval/models/sale_order.py:
```diff
@@ -0,0 +1,60 @@
+    )
+
+
+class SaleOrder(models.Model):
```
Please put one model per file.

rousseldenis commented on sale_product_approval/models/sale_order.py:
```diff
@@ -0,0 +1,60 @@
+
+    @api.depends("order_line.approved_sale")
+    def _compute_exceptions(self):
+        self.exceptions_sale_approval = any(
```
You should modify this in order to support multi recordset

rousseldenis commented on sale_product_approval/tests/test_product_approval.py:
```diff
@@ -0,0 +1,50 @@
+from odoo.tests.common import TransactionCase
+
+
+class TestSaleOrderLineDates(TransactionCase):
```
SavepointCase (until v14) is better in terms of performance.

You can look at this at 6:22:00 https://www.youtube.com/watch?v=lT___hG110Q

rousseldenis reviewed (changes requested):
Code review

rousseldenis commented:
@Chandresh-OSI Could you attend comments and solve Travis ?

As you break standard behaviour you need to find a mean to activate or not this behaviour.

rousseldenis commented:
@Chandresh-OSI @dreispt What's the status of this ?

rousseldenis commented:
@Chandresh-OSI @dreispt

dreispt committed:
[REF] sale_coupon_order_discount_fast_change_ui: fix pre-commit

dreispt commented:
All review comments addressed, merging.

/ocabot merge nobump

dreispt commented:
/ocabot merge nobump

dreispt committed:
[REF] sale_stock_picking_blocking: mark as rebel

dreispt commented:
/ocabot merge nobump
